api/classes/class-maybe-show-campaign.php

```diff
@@ -20,148 +20,94 @@ class Maybe_Show_Campaign extends Lightweight_API {
+
+		// Handle referer-based conditions.
+		$referer_url = filter_input( INPUT_SERVER, 'HTTP_REFERER', FILTER_SANITIZE_STRING );
+		if ( ! empty( $referer_url ) ) {
+			// Suppressing based on UTM Source parameter in the URL.
+			$utm_suppression = ! empty( $campaign->utm ) ? urldecode( $campaign->utm ) : null;
+			if ( $utm_suppression && stripos( urldecode( $referer_url ), 'utm_source=' . $utm_suppression ) ) {
+				$should_display                    = false;
+				$campaign_data['suppress_forever'] = true;
+			}
+
+			// Suppressing based on UTM Medium parameter in the URL.
+			$has_utm_medium_in_url = stripos( $referer_url, 'utm_medium=email' );
+			if (
+				$has_utm_medium_in_url &&
+				$settings->suppress_newsletter_campaigns &&
+				$has_newsletter_prompt
+			) {
+				$should_display                    = false;
+				$campaign_data['suppress_forever'] = true;
+			}
 		}
-		return false;
-	}
 
-	/**
-	 * Handle utm_medium query param, update transients if needed.
-	 *
-	 * @return bool Should campaign be shown.
-	 */
-	public function should_suppress_because_utm_medium() {
-		// Suppressing based on UTM Medium parameter in the URL. If:
-		// - the visitor came from email,
-		// - the suppress_newsletter_campaigns setting is on,
-		// - the pop-up has a newsletter form,
-		// then it should not be displayed.
-		$has_utm_medium_in_url     = stripos( $this->referer_url, 'utm_medium=email' );
-		$utm_medium_transient_name = $this->get_suppression_data_transient_name( 'utm_medium' );
+		$client_data                        = $this->get_client_data( $client_id );
+		$has_suppressed_newsletter_campaign = $client_data['suppressed_newsletter_campaign'];
 
+		// Handle suppressing a newsletter campaign if any newsletter campaign was dismissed.
 		if (
-			$this->suppress_newsletter_campaigns &&
-			$this->has_newsletter_prompt &&
-			( $has_utm_medium_in_url || $this->get_transient( $utm_medium_transient_name ) )
+			$has_newsletter_prompt &&
+			$settings->suppress_all_newsletter_campaigns_if_one_dismissed &&
+			$has_suppressed_newsletter_campaign
 		) {
-			$this->set_transient( $utm_medium_transient_name, true );
-			return true;
+			$should_display                    = false;
+			$campaign_data['suppress_forever'] = true;
 		}
-		return false;
-	}
 
-	/**
-	 * Should campaign be suppressed because it is a Newsletter campaign and a Newsletter campaign was previously dismissed.
-	 *
-	 * @return bool Should campaign be shown.
-	 */
-	public function should_suppress_because_newsletter_campaign_dismissed() {
-		// Suppressing a newsletter campaign if any newsletter campaign was dismissed.
-		$name = $this->legacy_get_suppression_data_transient_name_reversed( 'newsletter-campaign-suppression' );
-
-		if ( $this->suppress_all_newsletter_campaigns_if_one_dismissed && $this->has_newsletter_prompt && $this->get_transient( $name ) ) {
-			return true;
-		}
-		return false;
-	}
+		$this->save_campaign_data( $client_id, $campaign->id, $campaign_data );
```

Comment:
This statement incurs a database write every time. It would be more efficient if we could call it only if `$campaign data` has changed.

Reply:
Done in 452700a7100567bc45fc25551dded99e58c5ee8a

Reply:
Perfect. Since the introduction of the lightweight API I've been testing by clicking around a site and observing the debugging info in the API response. Initially you see multiple database reads and writes, then in subsequent requests these should drop to zero as the client info is stored in and retrieved from the cache. Before this change there was always a write for every campaign on the page, now the writes quickly resolve to 0.